Add redis DB and connector

return42 opened the pull request:
## What does this PR do?

### add redis connector

The redis connector is defined in the `settings.yml`:

    redis:
      url: ...

The default DB connector is a socket at::

    unix:///usr/local/searxng-redis/run/redis.sock?db=0

### script to build & install redis DB

A script to build & install a simple & isolated redis service, dedicated to SearXNG and connected via Unix socket.

    $ ./manage redis.help
    redis.:
      devpkg    : install essential packages to compile redis
      build     : build redis binaries at /800GBPCIex4/share/SearXNG/dist/redis/6.2.6/amd64
      install   : create user (searxng-redis) and install systemd service (searxng-redis)
      remove    : delete user (searxng-redis) and remove service (searxng-redis)
      shell     : start bash interpreter from user searxng-redis
      src       : clone redis source code to <path> and checkput 6.2.6
      useradd   : create user (searxng-redis) at /usr/local/searxng-redis
      userdel   : delete user (searxng-redis)
      addgrp    : add <user> to group (searxng-redis)
      rmgrp     : remove <user> from group (searxng-redis)

## How to test this PR locally?

To set up a redis instance simply use::

    $ ./manage redis.build
    $ sudo -H ./manage redis.install

To get access rights to this instance, your developer account needs to be added to the *searxng-redis* group::

    $ sudo -H ./manage redis.addgrp "${USER}"
    # don't forget to logout & login to get member of group

A redis DB connect can be tested by:

    >>> from searx.shared import redisdb
    >>> redisdb.init()
    True
    >>> db = redisdb.client()
    >>> db.set("foo", "bar")
    True
    >>> db.get("foo")
    b'bar'



## Related issues

- https://github.com/searxng/searxng/pull/618

return42 committed:
[mod] script to build & install a redis instance

A script to build & install a simple & isolated redis service, dedicated to
SearXNG and connected via Unix socket.

    $ ./manage redis.help
    redis.:
      devpkg    : install essential packages to compile redis
      build     : build redis binaries at /800GBPCIex4/share/SearXNG/dist/redis/6.2.6/amd64
      install   : create user (searxng-redis) and install systemd service (searxng-redis)
      remove    : delete user (searxng-redis) and remove service (searxng-redis)
      shell     : start bash interpreter from user searxng-redis
      src       : clone redis source code to <path> and checkput 6.2.6
      useradd   : create user (searxng-redis) at /usr/local/searxng-redis
      userdel   : delete user (searxng-redis)
      addgrp    : add <user> to group (searxng-redis)
      rmgrp     : remove <user> from group (searxng-redis)

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] test.shell: add utils/lib_redis.sh to shellcheck procedure

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[enh] add redis connector searx/shared/redisdb.py

Add a redis connector, the default DB connector is a socket at::

    unix:///usr/local/searxng-redis/run/redis.sock?db=0

To set up a redis instance simply use::

    $ ./manage redis.build
    $ sudo -H ./manage redis.install

A hint for developers:

To get access rights to this instance, your developer account needs to be added
to the *searxng-redis* group::

    $ sudo -H ./manage redis.addgrp "${USER}"
    # don't forget to logout & login to get member of group

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/shared/redisdb.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def client():
+    return redis.Redis.from_url(get_setting('redis.url'))
```
What about this:
```python
_client = None

def client():
    global _client
    if _client:
        return _client
    # not thread safe: in the worst case scenario,  two or more clients are initialized
    # only one is kept, the others are garbage collected.
    _client = redis.Redis.from_url(get_setting('redis.url'))
    return _client
```
?

So each call to the `client()` skip the overhead of the connection initialization.

AFIU, the client is [thread safe](https://github.com/redis/redis-py#thread-safety) and there is a [connection pool](https://github.com/redis/redis-py#connection-pools).

return42 replied:
> AFIU, the client is thread safe and there is a connection pool.

`redis.Redis.from_url` just wraps `ConnectionPool.from_url` ..

https://github.com/redis/redis-py/blob/15f315a496c3267c8cbcc6d6d9c6005ea4d4a4d5/redis/client.py#L852-L853

> So each call to the `client()` skip the overhead of the connection initialization.

AFAIK there is no need for, since `redis.Redis.from_url` just wraps `ConnectionPool.from_url` all the connection management and garbage collection is done by the ConnectionPool.

dalf replied:
`ConnectionPool.from_url` creates a new `ConnectionPool` at each call:
https://github.com/redis/redis-py/blob/15f315a496c3267c8cbcc6d6d9c6005ea4d4a4d5/redis/connection.py#L1200-L1206

I don't see any cache here.

return42 replied:
Hm, seems you are right .. I will have a look ..

return42 replied:
@dalf: thanks for your hint / fixed in dca83944

return42 committed:
[fix] redis: don't create a new connection at each client() call

Suggested-by: @dalf https://github.com/searxng/searxng/pull/686#pullrequestreview-844942973
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on utils/lib_redis.sh:
```diff
@@ -0,0 +1,348 @@
+    echo "${ARCH}"
+}
+
+# TODO: move this to the right place ..
```
where?

return42 replied:
> where?

`lib.sh` .. On the long term I will split (reorganize) the `lib.sh` in modules (we talked about a few month ago) .. ATM a func name `bash.cmd` does not fit good in the `lib.sh` ... ATM we have more urgent tasks, but I will an eye on.

dalf commented on searx/shared/redisdb.py:
```diff
@@ -0,0 +1,47 @@
+    return _client
+
+
+def init():
```
Where this function is called?

return42 replied:
We do not have a redis app right now, so it is not called.  Since redis DB is optional the intention is something like this:

```python
if redisdb.init():
   # read / write DB
else:
   # no DB available
```
